### ISSUE TITLE
High CPU usage while playing local tracks

### ISSUE THREAD

**Issue by pgzh** [jcorporation/myMPD#975]
I noticed relatively high CPU usage of chromium displaying the web ui of mympd on a Raspberry Pi 3B+ while a local track is playing.
mympd and chromium run on the same machine.
I'm running current Raspbian / Raspberry Pi OS with mympd v10.2.2 .

Observation 1: Chromium uses 60-80% CPU while playing a local track.
<code>top - 19:45:06 up 9 min,  2 users,  load average: 0.66, 0.85, 0.58
Tasks: 167 total,   3 running, 164 sleeping,   0 stopped,   0 zombie
%Cpu0  : 34.2 us, 12.0 sy,  0.0 ni, 53.4 id,  0.0 wa,  0.0 hi,  0.3 si,  0.0 st
%Cpu1  : 29.3 us,  9.2 sy,  0.0 ni, 61.6 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
%Cpu2  : 32.3 us,  9.9 sy,  0.0 ni, 57.8 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
%Cpu3  : 32.0 us, 10.3 sy,  0.0 ni, 57.7 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
MiB Mem :    909.6 total,    123.1 free,    330.1 used,    456.5 buff/cache
MiB Swap:    100.0 total,    100.0 free,      0.0 used.    504.9 avail Mem 

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND                                                                                                                                               
    867 pgzh      20   0  170.6g 140932 100344 R  79.6  15.1   3:52.52 chromium                                                                                                                                              
    809 pgzh      20   0   32.5g 118012  79996 R  59.9  12.7   2:25.50 chromium
</code>

Switching between home/playback/queue screens does not seem to make any difference.

Observation 2: As soon as the chromium window loses focus the CPU usage drops to virtually 0.
<code>top - 19:46:46 up 11 min,  2 users,  load average: 0.97, 1.02, 0.67
Tasks: 169 total,   1 running, 168 sleeping,   0 stopped,   0 zombie
%Cpu0  :  3.9 us,  0.3 sy,  0.0 ni, 95.8 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
%Cpu1  :  1.3 us,  0.7 sy,  0.0 ni, 98.0 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
%Cpu2  :  1.3 us,  2.6 sy,  0.0 ni, 96.1 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
%Cpu3  :  0.0 us,  0.3 sy,  0.0 ni, 99.7 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
MiB Mem :    909.6 total,     99.4 free,    357.1 used,    453.2 buff/cache
MiB Swap:    100.0 total,    100.0 free,      0.0 used.    481.9 avail Mem 

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND                                                                                                                                               
    491 mpd       20   0  761036  65596  33812 S   2.0   7.0   0:18.14 mpd                                                                                                                                                   
    779 pgzh      20   0   10172   3472   2656 R   1.3   0.4   0:06.89 top                                                                                                                                                   
    285 root       0 -20       0      0      0 I   0.3   0.0   0:06.20 kworker/u9:1-brcmf_wq/mmc1:0001:1                                                                                                                     
    546 61876     20   0  230644  11432   5100 S   0.3   1.2   0:04.35 mympd
</code>

Observation 3: Playing a webradio stream also does not cause high CPU usage of chromium, even when the chromium window is on top and has focus.
<code>top - 19:48:32 up 13 min,  2 users,  load average: 0.71, 0.95, 0.69
Tasks: 166 total,   1 running, 165 sleeping,   0 stopped,   0 zombie
%Cpu0  :  2.0 us,  0.3 sy,  0.0 ni, 97.7 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
%Cpu1  :  0.3 us,  1.3 sy,  0.0 ni, 98.3 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
%Cpu2  :  0.0 us,  0.7 sy,  0.0 ni, 99.0 id,  0.0 wa,  0.0 hi,  0.3 si,  0.0 st
%Cpu3  :  1.3 us,  1.3 sy,  0.0 ni, 97.3 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
MiB Mem :    909.6 total,    143.1 free,    312.7 used,    453.8 buff/cache
MiB Swap:    100.0 total,    100.0 free,      0.0 used.    526.2 avail Mem 

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND                                                                                                                                               
    491 mpd       20   0  742480  47056  33808 S   3.6   5.1   0:20.90 mpd                                                                                                                                                   
    133 root      20   0   49124  16416  15424 S   1.6   1.8   0:01.26 systemd-journal                                                                                                                                       
    398 root      20   0  221212   3468   2964 S   1.0   0.4   0:00.37 rsyslogd                                                                                                                                              
    779 pgzh      20   0   10172   3472   2656 R   1.0   0.4   0:07.97 top
</code>

Observation 4: The same behaviour can be observed for chromium running on a different machine connected to the mympd webserver on the RPi.
I get two chromium threads with about 25% CPU usage on a Intel Core i7 5775C while playing a local track, but virtually no CPU load when playing a webstream.

Is it possible the web ui is fetching the current playback position/information and updates the playback position at the bottom of the screen with a relatively high frequency [in a loop]?
Would it be possible to decrease the "current status update"-frequency or make it configurable to reduce CPU load?

I know web browser tend to consume an insane amount of resources, but since it only happens under certain circumstances it seems like too much CPU usage to me and I hope this can be improved.

As always, please let me know what else I can test or what information I can provide.

**Comment by jcorporation** [comment]
> Is it possible the web ui is fetching the current playback position/information and updates the playback position at the bottom of the screen with a relatively high frequency [in a loop]?
Would it be possible to decrease the "current status update"-frequency or make it configurable to reduce CPU load?

The setCounter function does this each second. It does not fetch any information, but only increases the progress bars and changes the counter text.

The difference between streams and playback of local files are the two progress bars (footer and queue view).

**Comment by jcorporation** [comment]
The load is in my environment lower as in your setup but much to high. It seems the high cpu usage results from the progress bar css transition. I quickly removed it in above commit and cpu usage in my environment dropped instantly. Can you check also?

**Comment by pgzh** [comment]
I will check and report back with new numbers on CPU usage on my system(s) tomorrow.
Thanks for the quick investigation and possible fix!

**Comment by pgzh** [comment]
> The load is in my environment lower as in your setup but much to high. It seems the high cpu usage results from the progress bar css transition. I quickly removed it in above commit and cpu usage in my environment dropped instantly. Can you check also?

I just built mympd after checking out the commit and can confirm that this fixed the issue for me.
To confirm I checked the system load with
<code>vmstat 1 10</code>
and averaged the system idle numbers ignoring the first in every run (since the first value was always significantly different from the rest, so I excluded the outlier for the average).

Scenarios as before:
1) Chromium not focused, mpd stopped
Before 99.6% idle
Now: 99.8% idle

2) Chromium focused, mpd stopped
Before: 96.5% idle
Now: 99.3% idle

3) Chromium focused, mpd playing local file
Before: 53.7% idle
Now: 98.7% idle [GREAT!! :-) ]

4) Chromium focused, mpd playing webstream
Before: 93.1% idle
Now: 99.7% idle


Since it's a RPi, the load average will probably be influenced by sporadic background activity a lot and I'd say this isn't any more accurate than +/- 2%, but the difference in case 3) is HUGE.

As far as I can tell you nailed this one, I have to admit I'm quite impressed how quickly you came up with that fix.
Thank you so much for investigating so quickly and your effort (again), I'm one more time looking forward to the next release containing the fix a lot! Keep up the great work!
